File: MCP_GATEWAY/SKill_SOCKET-Mcp_gateway/agents/SkillMatchAgent.js
const { findComplementaryUsers, testConnection } = require('../tools/dbTool');

class SkillMatchAgent {
  constructor(llmClient, toolRegistry) {
    this.llmClient = llmClient;
  }

  async run(query) {
    console.log(`SkillMatch Agent started with query: "${query}"`);
    
    try {
      // Test backend API connection first
      console.log('🔌 Testing backend API connection...');
      const connectionTest = await testConnection();
      if (!connectionTest.success) {
        throw new Error(`Backend API connection failed: ${connectionTest.error}`);
      }
      console.log(`✅ Backend API is accessible!`);
      
      // Extract skills from the query using LLM with better prompting
      const extractionPrompt = `
        Analyze this user query and extract the skills they need to learn and the skills they can offer to teach others.
        Return ONLY a valid JSON object with two arrays.
        
        Query: "${query}"
        
        Instructions:
        - "skillsRequired" = skills the user wants to learn, need, or is looking for
        - "skillsOffered" = skills the user can teach, offer, knows, or is good at
        - Look for phrases like: "I offer", "I can teach", "I know", "I'm good at", "I have experience in" (for offered skills)
        - Look for phrases like: "I need", "I want to learn", "I'm looking for", "help me with", "teach me" (for required skills)
        - Use common skill names like "JavaScript", "Python", "React", "Flutter", "Java", "Machine Learning", "Node.js", "CSS", "HTML", etc.
        - If user mentions wanting to exchange or trade skills, extract both sides appropriately
        - If no specific skills are mentioned, return empty arrays
        
        Example formats:
        Query: "I offer Flutter and need Java" → {"skillsRequired": ["Java"], "skillsOffered": ["Flutter"]}
        Query: "I can teach Python, want to learn React" → {"skillsRequired": ["React"], "skillsOffered": ["Python"]}
        Query: "Help me find users" → {"skillsRequired": [], "skillsOffered": []}
        
        Return only the JSON, no explanation:
      `;
      
      const extractionResult = await this.llmClient.generateText(extractionPrompt, 0.1);
      console.log("Raw LLM extraction result:", extractionResult);
      
      // Better JSON parsing with multiple attempts
      let skills;
      try {
        // First try: direct JSON parse
        skills = JSON.parse(extractionResult.trim());
      } catch (e1) {
        try {
          // Second try: extract JSON from text
          const jsonMatch = extractionResult.match(/\{[\s\S]*\}/);
          if (jsonMatch) {
            skills = JSON.parse(jsonMatch[0]);
          } else {
            throw new Error("No JSON found in response");
          }
        } catch (e2) {
          // Fallback: manual extraction
          console.log("JSON parsing failed, using fallback extraction");
          skills = this.extractSkillsFallback(query);
        }
      }
      
      // Validate and clean skills
      if (!skills.skillsRequired) skills.skillsRequired = [];
      if (!skills.skillsOffered) skills.skillsOffered = [];
      
      console.log("Extracted skills:", skills);
      
      // Find complementary users from database with timeout
      console.log('Searching for matching users...');
      const matchedUsers = await Promise.race([
        findComplementaryUsers(skills.skillsRequired, skills.skillsOffered),
        new Promise((_, reject) => setTimeout(() => reject(new Error('User search timeout after 8 seconds')), 8000))
      ]);
      console.log(`Found ${matchedUsers.length} matching users`);
      
      // Generate a response using the LLM
      let response;
      
      if (matchedUsers.length > 0) {
        const usersData = matchedUsers.map((user, index) => 
          `${index + 1}. **${user.name || user.firstName + ' ' + user.lastName || 'User'}** (${user.email})
   • 💡 **Offers**: ${user.skillsOffered.length > 0 ? user.skillsOffered.join(', ') : 'Not specified'}
   • 🎯 **Needs**: ${user.skillsRequired.length > 0 ? user.skillsRequired.join(', ') : 'Not specified'}
   • 📍 **Location**: ${user.location || 'Not specified'}`
        ).join('\n\n');
        
        response = `🎯 **Perfect! I found ${matchedUsers.length} user${matchedUsers.length > 1 ? 's' : ''} with complementary skills:**

**Your Profile:**
• 📚 **Want to learn**: ${skills.skillsRequired.join(', ') || 'Not specified'}
• 💡 **Can teach**: ${skills.skillsOffered.join(', ') || 'Not specified'}

**Recommended Connections:**

${usersData}

💡 **Why these matches work:**
These users offer skills you want to learn and need skills you can teach - creating perfect skill exchange opportunities!

**Next Steps:**
1. Reach out to these users through the platform
2. Propose a skill exchange arrangement
3. Set up learning sessions or mentorship
4. Build lasting professional connections`;
      } else {
        response = `🔍 **No exact matches found, but don't worry!**

**Your Skills Profile:**
• 📚 **Want to learn**: ${skills.skillsRequired.join(', ') || 'Not specified'}
• 💡 **Can teach**: ${skills.skillsOffered.join(', ') || 'Not specified'}

**Suggestions to find skill partners:**

1. **Broaden your search**: Consider related skills (e.g., if you want React, look for JavaScript experts)
2. **Update your profile**: Make sure your skills are clearly listed
3. **Be proactive**: Reach out to users with similar interests
4. **Join communities**: Look for study groups or skill-sharing circles
5. **Check back later**: New users join regularly!

**Alternative approach**: Consider offering your skills in exchange for general mentorship or project collaboration opportunities.`;
      }
      
      return {
        matches: matchedUsers,
        response: response,
        query: {
          skillsRequired: skills.skillsRequired,
          skillsOffered: skills.skillsOffered
        },
        matchCount: matchedUsers.length
      };
    } catch (error) {
      console.error("Error in SkillMatch Agent:", error);
      
      // Provide more specific error messages based on error type
      let errorMessage = error.message;
      let suggestions = [
        "1. Check your internet connection",
        "2. Try again in a few moments",
        "3. Use simpler skill names like 'JavaScript', 'Python', 'React'"
      ];
      
      if (error.message.includes('timeout') || error.message.includes('buffering')) {
        errorMessage = "Database connection timed out";
        suggestions = [
          "1. The database server might be slow or unavailable",
          "2. Check your internet connection",
          "3. Try again in a few minutes",
          "4. Contact support if the issue persists"
        ];
      } else if (error.message.includes('connection')) {
        errorMessage = "Could not connect to the database";
        suggestions = [
          "1. Check if the database service is running",
          "2. Verify your internet connection",
          "3. Try again later"
        ];
      }
      
      return {
        error: error.message,
        response: `❌ **Database Connection Issue**

**Problem**: ${errorMessage}

**What you can try:**
${suggestions.map(s => `• ${s}`).join('\n')}

**Alternative**: While we fix this, try asking general questions about skills or learning paths!`,
        matches: [],
        matchCount: 0
      };
    }
  }

  // Fallback method to extract skills when JSON parsing fails
  extractSkillsFallback(query) {
    const lowerQuery = query.toLowerCase();
    const commonSkills = [
      'javascript', 'python', 'java', 'react', 'flutter', 'dart', 'node.js', 'nodejs',
      'angular', 'vue', 'typescript', 'c++', 'c#', 'php', 'ruby', 'go', 'kotlin',
      'swift', 'html', 'css', 'sql', 'mongodb', 'mysql', 'postgresql', 'docker',
      'kubernetes', 'aws', 'azure', 'git', 'linux', 'machine learning', 'ai',
      'data science', 'spring boot', 'django', 'express', 'laravel'
    ];

    const skillsRequired = [];
    const skillsOffered = [];

    // Look for "learn", "need", "want" patterns
    const learnPatterns = ['learn', 'need', 'want to learn', 'studying', 'learning'];
    const teachPatterns = ['teach', 'offer', 'know', 'expert in', 'can help with'];

    learnPatterns.forEach(pattern => {
      if (lowerQuery.includes(pattern)) {
        commonSkills.forEach(skill => {
          if (lowerQuery.includes(skill) && !skillsRequired.includes(skill)) {
            skillsRequired.push(skill);
          }
        });
      }
    });

    teachPatterns.forEach(pattern => {
      if (lowerQuery.includes(pattern)) {
        commonSkills.forEach(skill => {
          if (lowerQuery.includes(skill) && !skillsOffered.includes(skill)) {
            skillsOffered.push(skill);
          }
        });
      }
    });

    // Default skills if none found
    if (skillsRequired.length === 0 && skillsOffered.length === 0) {
      if (lowerQuery.includes('flutter')) skillsRequired.push('Flutter');
      if (lowerQuery.includes('java')) skillsOffered.push('Java');
    }

    return { skillsRequired, skillsOffered };
  }
}

module.exports = SkillMatchAgent;

File: MCP_GATEWAY/SKill_SOCKET-Mcp_gateway/agents/skill_match_agent.py
import json
import re
import asyncio

from tools.db_tool import find_complementary_users, test_connection  # type: ignore[import-not-found]


class SkillMatchAgent:
    """Finds users with complementary skills for peer-to-peer skill exchange."""

    COMMON_SKILLS = [
        "javascript", "python", "java", "react", "flutter", "dart", "node.js",
        "nodejs", "angular", "vue", "typescript", "c++", "c#", "php", "ruby",
        "go", "kotlin", "swift", "html", "css", "sql", "mongodb", "mysql",
        "postgresql", "docker", "kubernetes", "aws", "azure", "git", "linux",
        "machine learning", "ai", "data science", "spring boot", "django",
        "express", "laravel",
    ]

    def __init__(self, llm_client, tool_registry):
        self.llm_client = llm_client

    async def run(self, query: str) -> dict:
        print(f'SkillMatch Agent started with query: "{query}"')

        try:
            # 1. Test API connection
            print("🔌 Testing backend API connection...")
            conn = await test_connection()
            if not conn.get("success"):
                raise RuntimeError(f"Backend API connection failed: {conn.get('error')}")
            print("✅ Backend API is accessible!")

            # 2. Extract skills using LLM
            extraction_prompt = self._build_extraction_prompt(query)
            raw = await self.llm_client.generate_text(extraction_prompt, 0.1)
            print(f"Raw LLM extraction result: {raw}")
            skills = self._parse_skills(raw, query)
            print(f"Extracted skills: {skills}")

            # 3. Search for complementary users (with 8 s timeout)
            print("Searching for matching users...")
            matched_users = await asyncio.wait_for(
                find_complementary_users(
                    skills.get("skillsRequired", []),
                    skills.get("skillsOffered", []),
                ),
                timeout=8.0,
            )
            print(f"Found {len(matched_users)} matching users")

            # 4. Build response
            response = self._format_response(skills, matched_users)
            return {
                "matches": matched_users,
                "response": response,
                "query": {
                    "skillsRequired": skills.get("skillsRequired", []),
                    "skillsOffered": skills.get("skillsOffered", []),
                },
                "matchCount": len(matched_users),
            }

        except Exception as exc:
            print(f"Error in SkillMatch Agent: {exc}")
            return self._error_response(str(exc))

    # ── helpers ───────────────────────────────────────────────────────────

    @staticmethod
    def _build_extraction_prompt(query: str) -> str:
        return (
            "Analyze this user query and extract the skills they need to learn "
            "and the skills they can offer to teach others.\n"
            "Return ONLY a valid JSON object with two arrays.\n\n"
            f'Query: "{query}"\n\n'
            "Instructions:\n"
            '- "skillsRequired" = skills the user wants to learn\n'
            '- "skillsOffered"  = skills the user can teach\n'
            "- Use common skill names like JavaScript, Python, React, Flutter, etc.\n\n"
            "Return only the JSON, no explanation:"
        )

    def _parse_skills(self, raw: str, query: str) -> dict:
        """Try JSON parsing, then regex, then fallback heuristic."""
        try:
            return json.loads(raw.strip())
        except json.JSONDecodeError:
            pass
        match = re.search(r"\{[\s\S]*\}", raw)
        if match:
            try:
                return json.loads(match.group(0))
            except json.JSONDecodeError:
                pass
        return self._extract_skills_fallback(query)

    def _extract_skills_fallback(self, query: str) -> dict:
        lower = query.lower()
        skills_required: list[str] = []
        skills_offered: list[str] = []

        learn_patterns = ["learn", "need", "want to learn", "studying", "learning"]
        teach_patterns = ["teach", "offer", "know", "expert in", "can help with"]

        for pattern in learn_patterns:
            if pattern in lower:
                for skill in self.COMMON_SKILLS:
                    if skill in lower and skill not in skills_required:
                        skills_required.append(skill)

        for pattern in teach_patterns:
            if pattern in lower:
                for skill in self.COMMON_SKILLS:
                    if skill in lower and skill not in skills_offered:
                        skills_offered.append(skill)

        if not skills_required and not skills_offered:
            if "flutter" in lower:
                skills_required.append("Flutter")
            if "java" in lower:
                skills_offered.append("Java")

        return {"skillsRequired": skills_required, "skillsOffered": skills_offered}

    @staticmethod
    def _format_response(skills: dict, matched_users: list[dict]) -> str:
        req = ", ".join(skills.get("skillsRequired", [])) or "Not specified"
        off = ", ".join(skills.get("skillsOffered", [])) or "Not specified"

        if matched_users:
            lines = []
            for i, u in enumerate(matched_users, 1):
                name = u.get("name") or f"{u.get('firstName', '')} {u.get('lastName', '')}".strip() or "User"
                offered = ", ".join(u.get("skillsOffered", [])) or "Not specified"
                needed = ", ".join(u.get("skillsRequired", [])) or "Not specified"
                loc = u.get("location", "Not specified")
                lines.append(
                    f"{i}. **{name}** ({u.get('email', '')})\n"
                    f"   • 💡 **Offers**: {offered}\n"
                    f"   • 🎯 **Needs**: {needed}\n"
                    f"   • 📍 **Location**: {loc}"
                )
            users_data = "\n\n".join(lines)
            n = len(matched_users)
            return (
                f"🎯 **Perfect! I found {n} user{'s' if n > 1 else ''} "
                f"with complementary skills:**\n\n"
                f"**Your Profile:**\n• 📚 **Want to learn**: {req}\n"
                f"• 💡 **Can teach**: {off}\n\n"
                f"**Recommended Connections:**\n\n{users_data}\n\n"
                "💡 **Why these matches work:**\n"
                "These users offer skills you want to learn and need skills "
                "you can teach — creating perfect skill exchange opportunities!\n\n"
                "**Next Steps:**\n"
                "1. Reach out to these users through the platform\n"
                "2. Propose a skill exchange arrangement\n"
                "3. Set up learning sessions or mentorship\n"
                "4. Build lasting professional connections"
            )

        return (
            f"🔍 **No exact matches found, but don't worry!**\n\n"
            f"**Your Skills Profile:**\n• 📚 **Want to learn**: {req}\n"
            f"• 💡 **Can teach**: {off}\n\n"
            "**Suggestions to find skill partners:**\n\n"
            "1. **Broaden your search**: Consider related skills\n"
            "2. **Update your profile**: Make sure your skills are clearly listed\n"
            "3. **Be proactive**: Reach out to users with similar interests\n"
            "4. **Join communities**: Look for study groups or skill-sharing circles\n"
            "5. **Check back later**: New users join regularly!"
        )

    @staticmethod
    def _error_response(message: str) -> dict:
        if "timeout" in message.lower() or "buffering" in message.lower():
            error_msg = "Database connection timed out"
            suggestions = [
                "The database server might be slow or unavailable",
                "Check your internet connection",
                "Try again in a few minutes",
                "Contact support if the issue persists",
            ]
        elif "connection" in message.lower():
            error_msg = "Could not connect to the database"
            suggestions = [
                "Check if the database service is running",
                "Verify your internet connection",
                "Try again later",
            ]
        else:
            error_msg = message
            suggestions = [
                "Check your internet connection",
                "Try again in a few moments",
                "Use simpler skill names like 'JavaScript', 'Python', 'React'",
            ]

        bullets = "\n".join(f"• {s}" for s in suggestions)
        return {
            "error": message,
            "response": (
                f"❌ **Database Connection Issue**\n\n"
                f"**Problem**: {error_msg}\n\n"
                f"**What you can try:**\n{bullets}\n\n"
                "**Alternative**: While we fix this, try asking general questions "
                "about skills or learning paths!"
            ),
            "matches": [],
            "matchCount": 0,
        }
